File: python/search_in_rotated_sorted_array.py
class Solution:
    def search(self, nums: list[int], target: int) -> int:
        """time complexity: O(log n)"""
        l: int = 0
        r: int = len(nums) - 1
        while l <= r:
            mid: int = (l + r) // 2
            if nums[mid] == target:
                return mid
            if nums[l] <= nums[mid]:
                if nums[l] <= target and target < nums[mid]:
                    r = mid - 1
                else:
                    l = mid + 1
            else:
                if nums[mid] < target and target <= nums[r]:
                    l = mid + 1
                else:
                    r = mid - 1
        return -1


def test_search_case_1():
    # arrange
    nums: list[int] = [4, 5, 6, 7, 0, 1, 2]
    target: int = 0
    expected: int = 4

    # act
    solution = Solution()
    actual = solution.search(nums, target)

    # assert
    assert expected == actual


def test_search_case_2():
    # arrange
    nums: list[int] = [4, 5, 6, 7, 0, 1, 2]
    target: int = 3
    expected: int = -1

    # act
    solution = Solution()
    actual = solution.search(nums, target)

    # assert
    assert expected == actual


def test_search_case_3():
    # arrange
    nums: list[int] = [1]
    target: int = 0
    expected: int = -1

    # act
    solution = Solution()
    actual = solution.search(nums, target)

    # assert
    assert expected == actual


def test_search_case_4():
    # arrange
    nums: list[int] = [5, 1, 3]
    target: int = 5
    expected: int = 0

    # act
    solution = Solution()
    actual = solution.search(nums, target)

    # assert
    assert expected == actual

File: rust/src/code/search_in_rotated_sorted_array.rs
pub struct Solution {}

impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut l = 0;
        let mut r = nums.len() - 1;

        while l <= r {
            let mid = (l + r) / 2;
            if nums[mid] == target {
                return mid as i32;
            }

            if nums[l] <= nums[mid] {
                // left part
                if nums[mid] < target || target < nums[l] {
                    l = mid + 1;
                } else {
                    r = mid - 1;
                }
            } else {
                // right part
                if nums[mid] > target || target > nums[r] {
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            }
        }

        return -1;
    }
}

#[cfg(test)]
mod test {
    use super::Solution;

    #[test]
    fn test_search_case_1() {
        let nums = vec![4, 5, 6, 7, 0, 1, 2];
        let target = 0;
        let expected = 4;
        let actual = Solution::search(nums, target);

        assert_eq!(actual, expected);

        let nums2 = vec![4, 5, 6, 7, 0, 1, 2];
        let target2 = 3;
        let expected2 = -1;
        let actual2 = Solution::search(nums2, target2);

        assert_eq!(actual2, expected2);
    }
}
